### PR TITLE
Adding a Cron Scaler to KEDA

### PR DESCRIPTION
**Pull request by ckuduvalli**
Added a Cron scaler in KEDA to scale based on Cron schedules.

### Checklist

- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [ ] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Tests have been added

Fixes #

### CODE REVIEW

**Comment by bharathguvvala** [comment]
are there no changes to scaledobject CRD in this feature?

**Comment by bharathguvvala** [comment]
Is there a doc which i can refer to regarding the LLD? I have been part of some discussions but I have no inkling on the final LLD that's agreed upon. In my mind I am still thinking that the schedule is injected as a time series, not sure if that's the case.

**Comment by ckuduvalli** [comment]
> are there no changes to scaledobject CRD in this feature?

@bharathguvvala There are no changes to the scaledObject CRD for any new scaler

**Comment by ckuduvalli** [comment]
Sample CRD

apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: cron-scaledobject
  namespace: default
spec:
  scaleTargetRef:
    deploymentName: nginx-deployment-basic
  pollingInterval: 30
  cooldownPeriod:  60
  minReplicaCount: 1
  maxReplicaCount: 100
  triggers:
  - type: cron
    metadata:
      timezone: Asia/Kolkata
      start: 57 18 18 May ?
      end: 00 19 18 May ?
      desiredReplicas: "10"
  - type: prometheus
    metadata:
      serverAddress: http://10.33.246.111:9090/
      metricName: cpu_usage
      threshold: "3"
      query: avg((sum(sum(rate(container_cpu_usage_seconds_total{pod=~"dart-deployment-.*"}[5m]))by(pod)/sum(container_spec_cpu_quota{pod=~"dart-deployment-.*"}/container_spec_cpu_period{pod=~"dart-deployment-.*"})by(pod))by(pod))*100)

**Comment by bharathguvvala** [comment]
https://github.com/kedacore/keda/blob/master/CREATE-NEW-SCALER.md

**Comment by bharathguvvala** [comment]
Based on our discussion, we need to capture when isActive() is called and getMetrics() is called only if the scaler is active.

File: pkg/handler/scale_handler.go [hunk 1]
```diff
@@ -228,7 +228,7 @@ func (h *ScaleHandler) GetDeploymentScalers(scaledObject *kedav1alpha1.ScaledObj
 			authParams["awsRoleArn"] = deployment.Spec.Template.ObjectMeta.Annotations[kedav1alpha1.PodIdentityAnnotationKiam]
 		}
 
-		scaler, err := h.getScaler(scaledObject.Name, scaledObject.Namespace, trigger.Type, resolvedEnv, trigger.Metadata, authParams, podIdentity)
+		scaler, err := h.getScaler(scaledObject, trigger.Type, resolvedEnv, trigger.Metadata, authParams, podIdentity)
```
Comment (bharathguvvala) [inline pkg/handler/scale_handler.go:231]: Can this refactoring be separated into a different PR?

File: pkg/scalers/cron_scaler.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+type cronScaler struct {
+	metadata               *cronMetadata
+	deploymentName         string
+	namespace              string
+	startCron              *cron.Cron
+	endCron                *cron.Cron
+	client                 client.Client
+}
+
+type cronMetadata struct {
```
Comment (bharathguvvala) [inline pkg/scalers/cron_scaler.go:36]: As discussed, we need to outline the upgrade plan in the event of this cronMetadata's schema changing. What would happen to old scaledobjects created with the older schema etc.

File: pkg/handler/scale_handler.go (hunk 1 above)
Comment (bharathguvvala) [reply pkg/handler/scale_handler.go:231]: Upon looking further into the PR, I get why this is done.

File: pkg/handler/scale_handler.go [hunk 3]
```diff
@@ -330,12 +330,14 @@ func (h *ScaleHandler) getScaler(name, namespace, triggerType string, resolvedEn
 		return scalers.NewAzureEventHubScaler(resolvedEnv, triggerMetadata)
 	case "prometheus":
 		return scalers.NewPrometheusScaler(resolvedEnv, triggerMetadata)
+	case "cron":
+		return scalers.NewCronScaler(h.client, scaledObject.Spec.ScaleTargetRef.DeploymentName, scaledObject.Namespace, resolvedEnv, triggerMetadata)
```
Comment (bharathguvvala) [inline pkg/handler/scale_handler.go:334]: Why's the constructor signature deviating from that of the other scalers?

Comment (ckuduvalli) [reply pkg/handler/scale_handler.go:334]: If you see externalScaler, that also has name and namespace. We would need it to get the deployment replicas.

File: pkg/scalers/cron_scaler.go (hunk 2 above)
Comment (bharathguvvala) [inline pkg/scalers/cron_scaler.go:36]: More appropriate to name it cronScalerMetadata

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:36]: this is true with any scaler. But we will have to internally have this plan for our clients

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:36]: I referred to other scalers and used the same convention. They haven't used the name scaler for metadata

File: pkg/scalers/cron_scaler.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	return &meta, nil
+}
+
+// IsActive checks if the startTime or endTime has reached
+func (s *cronScaler) IsActive(ctx context.Context) (bool, error) {
+	startTime := s.startCron.Entries()[0].Next.Unix()
+	endTime := s.endCron.Entries()[0].Next.Unix()
+	currentTime := time.Now().Unix()
+
+    if startTime < endTime && currentTime < startTime {
```
Comment (bharathguvvala) [inline pkg/scalers/cron_scaler.go:130]: We can compound first expression and the last one for brevity.

File: pkg/scalers/cron_scaler.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	return nil
+}
+
+// GetMetricSpecForScaling returns the metric spec for the HPA
+func (s *cronScaler) GetMetricSpecForScaling() []v2beta1.MetricSpec {
+    return []v2beta1.MetricSpec{
+		{
+			External: &v2beta1.ExternalMetricSource{
+				MetricName:         cronMetricName,
+				TargetAverageValue: resource.NewQuantity(int64(defaultDesiredReplicas), resource.DecimalSI),
```
Comment (bharathguvvala) [inline pkg/scalers/cron_scaler.go:151]: We should explicitly use 1 imo instead of defaultDesiredReplicas. What if the defaultDesiredReplicas value changes to something else.

Comment (bharathguvvala) [reply pkg/scalers/cron_scaler.go:151]: even though defaultDesiredReplicas is 1 here when the variable is used in this context it confuses the readers.

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:151]: ok. will check again

File: pkg/scalers/cron_scaler.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	defaultDesiredReplicas = 1
+	cronMetricName         = "ReplicaCount"
+	cronMetricType         = "External"
+)
+
+type cronScaler struct {
+	metadata               *cronMetadata
+	deploymentName         string
+	namespace              string
+	startCron              *cron.Cron
```
Comment (bharathguvvala) [inline pkg/scalers/cron_scaler.go:31]: seems like these startCron and endCron and redundant. you can instead do this as part of isActive() just get s.metadata.start and s.metadata.end and evaluate the next instance of cron run and use it. The scalers are anyway transient and are only instantiated per every isActive and getMetrics request, so starting the cron and storing them in startCron and endCron doesn't make sense.

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:31]: but the cron has to be started only once. It should not be started every time isActive method is called. Unless the cron is started, the next instance will be  returned as zero which is what I tested.

Comment (bharathguvvala) [reply pkg/scalers/cron_scaler.go:31]: the scaler is instantiated for every isActive and getMetrics request and they are not persistent so this effectively means the cron is being started every time.

Comment (bharathguvvala) [reply pkg/scalers/cron_scaler.go:31]: image

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:31]: true. my point was unless cron is started, I won't be able to get the next instant. But I see your point. Let me relook

Comment (bharathguvvala) [reply pkg/scalers/cron_scaler.go:31]: yea you'll have to start the cron and it can be done inside the isActive()

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:26]: done

Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:26]: done

File: pkg/scalers/cron_scaler_test.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	{map[string]string{}, true},
+	{validCronMetadata, false},
+	{map[string]string{"timezone": "Asia/Kolkata", "start": "30 * * * *", "end": "45 * * * *"}, true},
+	{map[string]string{"start": "30 * * * *", "end": "45 * * * *", "desiredReplicas": "10"}, true},
+}
+
+var tz, _ = time.LoadLocation(validCronMetadata["timezone"])
+var currentDay = time.Now().In(tz).Weekday().String()
+
+func TestCronParseMetadata(t *testing.T) {
```
Comment (Purushotham233) [inline pkg/scalers/cron_scaler_test.go:33]: Am not sure about the test cases for private methods.  Behaviour test cases(i.e., various scenarios triggering all the branches and private method) are better rather than just unit tests.

Problem with unit tests they may need changes or become irrelevant when implementation changes. Whereas Behaviour test wont have that concern.

File: pkg/scalers/cron_scaler.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+			Type: cronMetricType,
+		},
+	}
+}
+
+// GetMetrics finds the current value of the metric
+func (s *cronScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+
+	var currentReplicas = int64(defaultDesiredReplicas)
+	isActive, _ := s.IsActive(ctx)
```
[the remark this answers is not in the document]
Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:148]: sure

File: pkg/scalers/cron_scaler.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+		return false, fmt.Errorf("error initializing start cron: %s", scronErr)
+	}
+
+	endTime, ecronErr := getCronTime(location, s.metadata.end)
+	if ecronErr != nil {
+		return false, fmt.Errorf("error intializing end cron: %s", ecronErr)
+	}
+
+	currentTime := time.Now().Unix()
+    if startTime < endTime && currentTime < startTime {
```
[the remark this answers is not in the document]
Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:117]: its not just a cron meta validation Purushotham. Let me give an example. Consider the start as 30 * * * * (every hour starting at 30th minute) and end as 45 * * * * (every hour starting at 45th minute). This IsActive gets called every few seconds (30s or so) depending on the polling interval. Now, Once the current time crosses the 30th minute, the next entry in the start cron will point to the 30th minute in the next hour. In that case currentTime will be indeed less than the startTime, but startTime > endTime. But scaler still needs to remain active. Hence, this check startTime < endTime.

File: pkg/scalers/cron_scaler.go (hunk 4 above)
Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go:117]: The last check ideally is not required. But for the method signature to work, I need to have that else part. Bcos I want to explicitly have the check currentTime <= endTime

File: pkg/scalers/cron_scaler_test.go (hunk 7 above)
Comment (ckuduvalli) [reply pkg/scalers/cron_scaler_test.go:33]: other scalers have mostly implemented unit testing only Puru. That said, I agree with you that there should be behaviour tests which should be part of the framework itself. As part of a separate pr, let me see if I can bring in those tests into the framework.

File: pkg/scalers/cron_scaler.go (hunk 5 above)
Comment (ckuduvalli) [reply pkg/scalers/cron_scaler.go]: have changed this.

File: pkg/handler/scale_handler.go (hunk 3 above)
Comment (ckuduvalli) [reply pkg/handler/scale_handler.go]: this is irrelevant now

File: pkg/scalers/cron_scaler.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
```
Comment (neerajb) [inline pkg/scalers/cron_scaler.go:117]: Formatting looks off. run go fmt.
